[ML] Refactors Field statistics table from an embeddable to a registered React component

qn895 opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/174965. This PR refactors Field statistics table from an embeddable to a async React component. 

**Context**
Initially when embeddable was implemented, we could not register dataVisualizer as an optional plugin due to the cyclical dependencies and enforced package import rules (from x-pack to discover). However, due to changes in the infrastructure in the last few years, we can now register x-pack plugins as a dependency. This eliminates the need for an embeddable as a wrapper around the field statistics table. 

**Reviewer's note**
- Discover team: `dataVisualizer` plugin is now added as an optional dependency in Discover
- Kibana Core team: `dataVisualizer` plugin is now added as an optional dependency in Discover, `discover` removed from Data visualizer plugin

https://github.com/elastic/kibana/assets/43350163/7800291a-ad8c-447e-9bfc-e4f279b05d6a


https://github.com/elastic/kibana/assets/43350163/d7422e11-a0c7-4a8b-8f47-783e453f02a8

Follow up after this PR:
- https://github.com/elastic/kibana/pull/180849
- https://github.com/elastic/kibana/issues/181904

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.14"]} ONMERGE-->

mgiota commented:
@qn895 I am linking your PR with this [one](https://github.com/elastic/kibana/pull/181392). For now, each embeddable that migrates to the new architecture, needs to add the `data-shared-item` attribute, so that Kibana reporting screenshot tool doesn't complain with timeout error.

qn895 commented:
/ci

qn895 commented:
/ci

qn895 commented:
/ci

qn895 commented:
/ci

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

davismcphee reviewed (commented):
The Discover changes are looking good overall! Just noticed a couple of issues when testing.

Changes don't seemed to be picked up by the field statistics tab when navigating through browser history in Discover:

https://github.com/elastic/kibana/assets/25592674/f0775b8a-2c6c-4f02-bb76-b72c875300e0

It looks like filters aren't being applied for field statistics panels in Dashboard. I used the `-` button to filter out the value, although I'm not sure if that makes a difference:
<img width="1809" alt="dashboard_filters" src="https://github.com/elastic/kibana/assets/25592674/76244c1d-e940-4286-b5ea-94ee568acb2d">

qn895 commented:
/ci

qn895 commented:
/ci

qn895 commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/embeddables/grid_embeddable/embeddable_field_stats_table.tsx:
```diff
@@ -20,24 +20,22 @@ import { EmbeddableNoResultsEmptyPrompt } from './embeddable_field_stats_no_resu
-  input: Required<DataVisualizerGridEmbeddableInput, 'dataView'>;
-  onOutputChange?: (ouput: any) => void;
-}) => {
+const EmbeddableFieldStatsTableWrapper = (
```
Fixed here https://github.com/elastic/kibana/pull/181502/commits/731fdcbac613f22ee62da754cc2babb76b5208b1

qn895 commented:
/ci

qn895 commented:
/ci

qn895 commented:
Thanks @davismcphee and @peteharverson for testing. Per @ThomThomson's recommendation, I have migrated the field statistics table off the embeddable architecture and instead simply things by exposing it as a React component that Discover can consume. I've updated the PR description with the additional context to the change. 

I've also fixed the issue with filters now. With the update, I think it works correctly now with the navigation via browser, but some additional testing would be appreciated. Also one thing to note is that for text fields (like `agent` field in Kibana sample data logs), it can only sample 1000 documents _maximum_. So if the dataset returns 1000+ hits, it will still only show 1000.



https://github.com/elastic/kibana/assets/43350163/c55d6a3f-1919-4b54-93ed-23d2181d3f34


https://github.com/elastic/kibana/assets/43350163/4f014826-ebbc-4a50-92bf-be7d0ad1de95


https://github.com/elastic/kibana/assets/43350163/da0bd647-0853-4a21-bf5d-4d2711580ac9

The saved search dashboard field statistics panel will also try to retain **both** the filters from the original saved search as well as the dashboard. 

https://github.com/elastic/kibana/assets/43350163/eaf8ea58-302b-4235-adc1-a23d1483d182

peteharverson commented:
I found this issue with filters and the browser history. If you add a filter, then edit it, hitting the 'back' button takes you to a state where you end up with the original and edited filter:


https://github.com/elastic/kibana/assets/7405507/9339da9e-0661-4235-91ac-82ad9ef9bc18

ThomThomson commented:
> I have migrated the field statistics table off the embeddable architecture and instead simply things by exposing it as a React component that Discover can consume.

This looks great! Thank you for doing this. One small nitpick is that some of the files / components still have `embeddable` in the name. If it isn't too much trouble I'd recommend removing that to avoid confusion.

Additionally, if you'd still like to place this on its own on a Dashboard in the future, feel free to reach out to @elastic/kibana-presentation - it should still be a straightforward process.

qn895 commented:
@elasticmachine merge upstream

davismcphee commented on src/plugins/discover/public/application/main/components/field_stats_table/field_stats_table.tsx:
```diff
@@ -50,163 +46,61 @@ export const FieldStatisticsTable = (props: FieldStatisticsTableProps) => {
-    showPreviewByDefault,
-  ]);
+  const lastReloadRequestTime$ = stateContainer?.dataState?.refetch$
+    ? stateContainer?.dataState?.refetch$.pipe(map(() => Date.now()))
```
Nit: since this creates a new observable on each render, I think it will resubscribe in `useObservable` every time.

qn895 replied:
Thanks for catching that. Updated here https://github.com/elastic/kibana/pull/181502/commits/e6603b563419e44e0d45bff8d663d14d86fa91cf.

davismcphee commented on src/plugins/discover/public/plugin.tsx:
```diff
@@ -180,6 +181,7 @@ export interface DiscoverSetupPlugins {
 export interface DiscoverStartPlugins {
   dataViews: DataViewsServicePublic;
   dataViewEditor: DataViewEditorStart;
+  dataVisualizer?: DataVisualizerPluginStart;
```
I'm curious if this needs to be optional? We now have conditional logic around the view toggle, etc. where we didn't before, but do we expect it to ever actually be undefined?

qn895 replied:
I think it's fine as an optional service since: 
1) User can already optionally disabled Field statistics in advanced settings or disable data_visualizer plugin in the yml config 
2) We want it so that the lack of dataVisualizer's service will not break the Discover's basic experience.

davismcphee replied:
Oh yeah, disabling the `data_visualizer` plugin would do it. Makes sense, thanks!

davismcphee reviewed (commented):
The Discover changes look good to me, and I confirmed the issues I was seeing before have now been resolved! The code is much simple in the field stats table component too, which is great.

I just have one question around some conditional logic I'd like clarification on before approving.

darnautov commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/embeddables/grid_embeddable/types.ts:
```diff
@@ -40,17 +84,46 @@ export interface DataVisualizerGridInput<T = Query> {
+  onTableUpdate?: (update: Partial<DataVisualizerTableState>) => void;
+}
+
+export type ESQLDataVisualizerGridEmbeddableState = Omit<
```
I see lots of types that contain `embeddable`. As it's no longer an embeddable, shall we rename the types?

qn895 replied:
We will have another follow up in 8.15 that makes this an embeddable to be used in dashboard. So I'd like to keep this as is for now to reduce churn.

peteharverson reviewed (approved):
Tested latest changes and LGTM

qn895 commented:
@elasticmachine merge upstream

davismcphee reviewed (approved):
Great work! The Data Discovery changes LGTM 👍

qn895 commented:
@elasticmachine merge upstream

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.14|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 181502
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
